File: src/static_object_tracking.py
import numpy as np
import rospy
from autocycle.msg import Object, ObjectList

height = 10000              # vertical dimension in millimeters
width = 20000               # horizontal dimension in millimeters
object_length = 1000		# object length

objects = []
heading = 0
time = 0

def new_object(obj):
	objects.append(obj)

def static_object_tracking():
	global objects, heading, time
	# Registers the Node with the Master
	rospy.init_node("static_object_tracking")

	# Waits for data getter service
	rospy.wait_for_service('get_data')

	# Creates the service client that will collect data
	data_getter = rospy.ServiceProxy("get_data", GetData)

	# Subscribes to the object topic
	rospy.Subscriber("cycle/objects", Object, new_object)

	# Publishes objects to path topic
	pub = rospy.Publisher("cycle/path", ObejctList, queue_size=1)

	heading = data_getter("heading").data
	time = data_getter("met").data

	while rospy.ok():
		# Picks up new objects
		rospy.spinOnce()

		# Collects data
		new_heading = data_getter("heading").data
		delta_angle = new_heading - heading
		heading = new_heading

		new_time = data_getter("met").data
		delta_time = new_time - time
		time = new_time
		
		speed = data_getter("vel").data

		theta = delta_angle
		c, s = np.cos(theta), np.sin(theta)
		rotation_matrix = np.array([[c, -s], [s, c]])
		distance = speed * delta_time
		new = []
		for obj in objects:
			rotated = rotation_matrix @ np.array([[obj.x1, obj.x2], [obj.z1, obj.z2]])
			new.append(Object(rotated[0,0], rotated[1,0] - distance, rotated[0,1], rotated[1,1] - distance))
		objects = list(filter(lambda x: x.z1 >= 0 or x.z2 >= 0, new))

	pub.publish(objects)

File: src/static_object_tracking.cpp
#include <ros/ros.h>
#include <chrono>
#include <cmath>
#include <std_msgs/Float32.h>
#include <autocycle_extras/ObjectList.h>
#include <autocycle_extras/Object.h>

float prev_heading = -1;
float heading = -1;
float velocity = -1;
std::vector<autocycle_extras::Object> ol;

void get_heading(const std_msgs::Float32 data){
    heading = data.data;
}

void get_velocity(const std_msgs::Float32 data){
    velocity = data.data;
}

void get_objects(const autocycle_extras::ObjectList data){
    ol.insert(ol.end(), data.obj_lst.begin(), data.obj_lst.end());
}

int main(int argc, char **argv){
    // Registers the Node with the Master
    ros::init(argc, argv, "static_object_tracking");
    ros::NodeHandle nh;

    // Creates subscriber that updates heading
    ros::Subscriber head_sub = nh.subscribe("sensors/heading", 1, &get_heading);

    // Creates subscriber that updates velocity
    ros::Subscriber vel_sub = nh.subscribe("sensors/vel", 1, &get_velocity);

    // Creates subscriber that gets latest object_list
    ros::Subscriber ol_sub = nh.subscribe("cycle/objects", 1, &get_objects);

    // Publishes objects to path topic
    ros::Publisher pub = nh.advertise<autocycle_extras::ObjectList>("cycle/object_frame", 1);

    // Initializing variables
    float delta_angle, delta_time, c, s, distance;
    std::vector<autocycle_extras::Object> new_objs;
    autocycle_extras::ObjectList to_pub;

    while (ol.size() == 0 || heading == -1 || velocity == -1){
        ros::spinOnce();
    }

    auto start = std::chrono::high_resolution_clock::now();
    while(ros::ok()){
        prev_heading = heading;
        ros::spinOnce();
        new_objs.clear();

        delta_angle = heading - prev_heading;

        auto end = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
        auto start = std::chrono::high_resolution_clock::now();
        delta_time = ((float) duration.count())/1000.0;

        c = cos(delta_angle);
        s = sin(delta_angle);
        distance = velocity * delta_time;

        for(auto & i : ol){
	    autocycle_extras::Object rotated;
            rotated.z1 = i.z1*c - i.x1*s;
            rotated.x1 = i.z1*s + i.x1*c - distance;
            rotated.z2 = i.z2*c - i.x2*s;
            rotated.x2 = i.z2*s + i.x2*c - distance;
            if(rotated.z1 < 0 && rotated.z2 < 0){
                continue;
            }
            new_objs.emplace_back(rotated);
        }
        ol.clear();
        ol = new_objs;
        to_pub.obj_lst = ol;
        pub.publish(to_pub);
	ol.clear(); // DELETE THIS LINE ONCE INTERSECTION IS WORKING
    }
}
